lib/Dialect/LWE/IR/NewLWEAttributes.td

```diff
@@ -0,0 +1,352 @@
+    defined by tuples $(z_1, \dots, z_N)$ such that $\overline{z_i} =
+    \overline{z_{N-i+1}}$. Note that this property holds because polynomial
+    evaluation commutes with complex conjugates, and the second half of the
+    roots of unity evaluate are complex conjugates of the first half. The
+    converse, that any such tuple with complex conjugate symmetry has an
+    inverse under $\tau$ with all real coefficients, makes $\tau$ is a
+    bijection onto $H$. $\tau$ and its inverse are explicitly computable as
+    discrete Fourier Transforms.
+
+    Because of the symmetry in canonical embedding for real polynomials, inputs
+    to this encoding can be represented as a list of $N/2$ complex points, with
+    the extra symmetric structure left implicit. $\tau_N$ and its inverse can
+    also be explicitly computed without need to expand the vectors to length
+    $N$.
+
+    The rounding step is required to invert the decoding because, while
+    cleartexts must be (implicitly) in the subspace $H$, they need not be the
+    output of $\tau_N$ for an _integer_ polynomial. The rounding step ensures
+    we can use integer polynomial plaintexts for the FHE operations. There are
+    multiple rounding mechanisms, and this attribute does not specify which is
+    used, because in theory two ciphertexts that have used different roundings
+    are still compatible, though they may have different noise growth patterns.
+
+    The scaling parameter $\Delta$ is specified by the `scaling_factor`, which
+    are applied coefficient-wise using the same semantics as the
+    `constant_coefficient_encoding`.
+
+    A typical flow for the CKKS scheme using this encoding would be to apply an
+    inverse FFT operation to invert the canonical embedding to be a polynomial
+    with real coefficients, then encrypt scale the resulting polynomial's
+    coefficients according to the scaling parameters, then round to get integer
+    coefficients.
+
+    Example:
+
+    ```
+    #canonical_encoding = #lwe.inverse_canonical_encoding<scaling_factor=10000>
+    ```
+  }];
+}
+
+def LWE_FullCRTPackingEncoding
+  : LWE_EncodingAttrWithScalingParam<"FullCRTPackingEncoding", "full_crt_packing_encoding"> {
+  let summary = "An attribute describing encoded GLWE plaintexts via CRT slots.";
+  let description = [{
+    This encoding maps a list of integers via the Chinese Remainder Theorem (CRT) into the plaintext space.
+
+    Given a ring with irreducible ideal polynomial `f(x)` and coefficient
+    modulus `q`, `f(x)` can be decomposed modulo `q` into a direct product of
+    lower-degree polynomials. This allows full SIMD-style homomorphic operations
+    across the slots formed from each factor.
+
+    This attribute restricts on full CRT packing, where the polynomial `f(x)`
+    splits completely (into linear factors) and the number of slots equals the
+    degree of `f(x)`. This happens when `q` is prime and `q = 1 mod n`.
```

Comment:
This suggests we need a verifier for the check on q, which is defined elsewhere.

Reply:
I added a verifier on the plaintext space that ensures that the ring and encoding attribute are compatible that way